blockchain/utxocache.go

```diff
@@ -0,0 +1,909 @@
+		// spent are always evicted since they are unlikely to be accessed again.
+		// Additionally, entries that are contained in a block with a height less
+		// than the eviction height are evicted.
+		if entry == nil || entry.IsSpent() ||
+			entry.BlockHeight() < int64(evictionHeight) {
+
+			// Remove the entry from the cache.
+			delete(c.entries, outpoint)
+
+			// Update the total entry size of the cache.
+			if entry != nil {
+				c.totalEntrySize -= entry.size()
+			}
+
+			continue
+		}
+
+		// If the entry wasn't removed from the cache, clear the modified and
+		// fresh flags since it has been updated in the database.
+		entry.state &^= utxoStateModified
+		entry.state &^= utxoStateFresh
+	}
+
+	// Update the last flush on the cache instance now that the flush has been
+	// completed.
+	c.lastFlushHash = *bestHash
+	c.lastFlushTime = time.Now()
+
+	// Update the last eviction height on the cache instance if we evicted just
+	// now.
+	if evictionHeight != 0 {
+		c.lastEvictionHeight = evictionHeight
+	}
+
+	// Log that the flush has been completed and indicate the updated memory usage
+	// as it will be reduced due to evicting entries above.
+	if logFlush {
+		remainingEntries := len(c.entries)
+		flushedEntries := preFlushNumEntries - remainingEntries
+		memUsage = c.totalSize()
+		memUsageMiB := float64(memUsage) / 1024 / 1024
+		memUsagePercent := float64(memUsage) / float64(c.maxSize) * 100
+		log.Debugf("UTXO cache flush completed (%d entries flushed, %d entries "+
+			"remaining, %.2f MiB (%.2f%%))", flushedEntries, remainingEntries,
+			memUsageMiB, memUsagePercent)
+	}
+
+	return nil
+}
+
+// MaybeFlush conditionally flushes the cache to the database.
+//
+// If the maximum size of the cache has been reached, or if the periodic flush
+// interval has been reached, then a flush is required.  Additionally, a flush
+// can be forced by setting the force flush parameter.
+//
+// This function is safe for concurrent access.
+func (c *UtxoCache) MaybeFlush(bestHash *chainhash.Hash, bestHeight uint32,
+	forceFlush bool, logFlush bool) error {
+
+	c.cacheLock.Lock()
+	if forceFlush || c.shouldFlush(bestHash) {
+		err := c.flush(bestHash, bestHeight, logFlush)
+		c.cacheLock.Unlock()
+		return err
+	}
+
+	c.cacheLock.Unlock()
+	return nil
+}
+
+// InitUtxoCache initializes the utxo cache by ensuring that the utxo set is
+// caught up to the tip of the best chain.
+//
+// Since the cache is only flushed to the database periodically, the utxo set
+// may not be caught up to the tip of the best chain.  This function catches the
+// utxo set up by replaying all blocks from the block after the block that was
+// last flushed to the tip block through the cache.
+//
+// This function should only be called during initialization.
+func (b *BlockChain) InitUtxoCache(tip *blockNode) error {
+	log.Infof("UTXO cache initializing (max size: %d MiB)...",
+		b.utxoCache.maxSize/1024/1024)
+	defer log.Info("UTXO cache initialization completed")
```

Comment:
Not complete if errored.

Reply:
Good point, updated to only print this log in the non-error paths.